qutebrowser/app.py

```diff
@@ -836,6 +837,30 @@ def __init__(self, args):
         self.focusObjectChanged.connect(self.on_focus_object_changed)
         self.setAttribute(Qt.AA_UseHighDpiPixmaps, True)
 
+        self.new_window.connect(self._on_new_window)
+
+    def _on_new_window(self, window):
+        window.tabbed_browser.shutting_down.connect(
+            functools.partial(self._on_window_closing, window)
+        )
+
+    def _on_window_closing(self, window):
+        self._undos.append(mainwindow.WindowUndoEntry(
+            geometry=window.saveGeometry(),
+            private=window.tabbed_browser.is_private,
+            tab_stack=window.tabbed_browser._undo_stack,
+        ))
+
+    def undo_last_window_close(self):
+        event = self._undos.pop()
+        window = mainwindow.MainWindow(
+            private=event.private,
+            geometry=event.geometry,
+        )
+        window.show()
+        window.tabbed_browser._undo_stack = event.tab_stack
```

Comment:
Not sure what to do about the private member access. I could add a `get_undo_entry()` to `MainWindow` but then mainwindow would still have to get them from the tabbedbrowser.